[JIT] Add support for backend-lowered submodules

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40842 [JIT] Add out-of-source-tree to_backend tests
* **#40841 [JIT] Add support for backend-lowered submodules**
* #40840 [JIT] Move TestBackend to test directory
* #40839 [JIT] Separate to_backend API into libtorch and libtorch_python

**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

Differential Revision: [D22418716](https://our.internmc.facebook.com/intern/diff/D22418716)

Commit:
[JIT] Add support for backend-lowered submodules

**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

suo commented on test/jit/test_backends.py:
```diff
@@ -69,47 +91,123 @@ def compare_py_jit_backend(self, name, input):
+        self.test_execution()
+
+
+class NestedModuleTest(JitBackendTestCase):
```
It's kind of hard to read this test because `NestedModule` and `MyModule` are defined at the top of this file. Generally we define any necessary test-specific fixtures inside the body of a test so that things stay local.

SplitInfinity replied:
It's hard to nest these because this test and the other test both use `MyModule`.

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -17,214 +17,220 @@ void initJitBackendBindings(PyObject* module) {
-        // This is a helper function for creating a new instance of the
-        // backend class.
-        static const auto create_backend_ct = CodeTemplate(R"(
+  auto codegen_lambda = [=](const std::string& backend_name,
```
Mind commenting on what the meaningful change here is? I think the indentation shift made GH freak out

SplitInfinity replied:
I wrapped all the codegen in a lambda so I could do:
```
return wrap_cpp_module(codegen_lambda(args)))
```

instead of 
```
return wrap_cpp_module([](...){
// lots
// of
// code
});
```

Commit:
Update on "[JIT] Add support for backend-lowered submodules"


**Summary**
This commit adds support for using `Modules` that have been lowered as
submodules in `ScriptModules`.

**Test Plan**
This commit adds execution and save/load tests to test_backends.py for
backend-lowered submodules.

**Fixes**
This commit fixes #40069.

[ghstack-poisoned]

suo reviewed (approved):
lgtm
